File: backend/controllers/userController.js
const supabase = require('../db');
const fs = require('fs');
const path = require('path');
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');
const { JWT_SECRET, BASE_URL } = require('../config/config');
const { uploadToVercelBlob } = require('../utils/vercelBlob');

// Helper function to determine the correct images directory path
const getImagesDirectory = (baseDir) => {
    console.log('Getting images directory from baseDir:', baseDir);
    
    // Try multiple possible paths for Vercel deployment
    let imagesDir = path.join(baseDir, '../../frontend/images');
    console.log('Trying imagesDir:', imagesDir);
    
    // If the standard path doesn't work, try alternative paths
    if (!fs.existsSync(path.join(baseDir, '../../frontend'))) {
        console.log('Standard frontend path not found, trying alternative paths');
        // Try without the extra ../
        imagesDir = path.join(baseDir, '../frontend/images');
        console.log('Trying alternative imagesDir:', imagesDir);
        
        // If that doesn't work, try with just frontend
        if (!fs.existsSync(path.join(baseDir, '../frontend'))) {
            imagesDir = path.join(baseDir, 'frontend/images');
            console.log('Trying another alternative imagesDir:', imagesDir);
        }
    }
    
    return imagesDir;
};

exports.getUserInfo = async (req, res) => {
    const userId = req.user.id;

    const { data: users, error } = await supabase
        .from('users')
        .select('id, firstname, middlename, lastname, email, profilePicture')
        .eq('id', userId);

    if (error) {
        console.error(error);
        return res.status(500).json({ success: false, message: 'An error occurred.' });
    }

    if (users.length > 0) {
        const user = users[0];
        // For Vercel Blob URLs, return as-is since they're already full URLs
        // For local images, return relative path
        if (user.profilePicture && !user.profilePicture.startsWith('http')) {
            user.profilePicture = user.profilePicture.replace(/\\/g, '/');
        }
        res.json({ success: true, user: user });
    } else {
        res.status(404).json({ success: false, message: 'User not found.' });
    }
};

exports.updateUserInfo = async (req, res) => {
    const userId = req.params.id;
    const { firstname, middlename, lastname, email } = req.body;

    if (req.user.id != userId) {
        return res.status(403).json({ success: false, message: 'Unauthorized to update this user.' });
    }

    if (!firstname || !lastname || !email) {
        return res.status(400).json({ success: false, message: 'First name, last name, and email are required.' });
    }

    const { data, error } = await supabase
        .from('users')
        .update({ firstname, middlename, lastname, email })
        .eq('id', userId);

    if (error) {
        console.error(error);
        return res.status(500).json({ success: false, message: 'Error updating user information.' });
    }

    if (data === null) {
        return res.status(404).json({ success: false, message: 'User not found or no changes made.' });
    }

    res.json({ success: true, message: 'Account information updated successfully!' });
};

exports.uploadProfilePicture = async (req, res) => {
    const userId = req.user.id;
    if (!req.file) {
        return res.status(400).json({ success: false, message: 'No file uploaded.' });
    }

    // Check if we're running on Vercel
    let profilepicturePath;
    if (process.env.VERCEL) {
        // Upload to Vercel Blob storage
        try {
            profilepicturePath = await uploadToVercelBlob(req.file, 'profiles');
            console.log('File uploaded to Vercel Blob:', profilepicturePath);
        } catch (uploadError) {
            console.error('Error uploading to Vercel Blob:', uploadError);
            return res.status(500).json({ success: false, message: 'Error uploading profile picture to Vercel Blob storage.' });
        }
    } else {
        // For local development, move file from /tmp to images directory
        if (req.file.path && req.file.path.startsWith('/tmp')) {
            // Move file from /tmp to images directory
            console.log('Current __dirname:', __dirname);
            const imagesDir = getImagesDirectory(__dirname);
            const targetPath = path.join(imagesDir, req.file.filename);
            console.log('Calculated targetPath:', targetPath);
            
            // Ensure the target directory exists
            console.log('Checking if directory exists:', imagesDir);
            if (!fs.existsSync(imagesDir)) {
                console.log('Creating directory:', imagesDir);
                try {
                    fs.mkdirSync(imagesDir, { recursive: true });
                    console.log('Directory created successfully');
                } catch (mkdirError) {
                    console.error('Error creating directory:', mkdirError);
                    return res.status(500).json({ success: false, message: 'Error creating image directory. Please try again or contact support.' });
                }
            } else {
                console.log('Directory already exists');
            }
            
            try {
                console.log('Moving file from', req.file.path, 'to', targetPath);
                fs.renameSync(req.file.path, targetPath);
                console.log('File moved successfully');
            } catch (moveError) {
                console.error('Error moving file:', moveError);
                // If rename fails, try copy and delete
                try {
                    console.log('Attempting to copy file instead');
                    fs.copyFileSync(req.file.path, targetPath);
                    fs.unlinkSync(req.file.path);
                    console.log('File copied and original deleted successfully');
                } catch (copyError) {
                    console.error('Error copying file:', copyError);
                    return res.status(500).json({ success: false, message: 'Error processing uploaded profile picture. Please try again or contact support.' });
                }
            }
        }
        profilepicturePath = `/images/${req.file.filename}`;
    }

    const { data, error } = await supabase
        .from('users')
        .update({ profilePicture: profilepicturePath })
        .eq('id', userId);

    if (error) {
        console.error('Error updating profile picture in DB:', error);
        // For Vercel deployments using Vercel Blob, we don't need to delete local files
        // For local development, delete the file
        if (!process.env.VERCEL) {
            let filePath = req.file.path;
            if (req.file.path && req.file.path.startsWith('/tmp')) {
                const imagesDir = getImagesDirectory(__dirname);
                filePath = path.join(imagesDir, req.file.filename);
            }
            fs.unlink(filePath, (unlinkErr) => {
                if (unlinkErr) console.error('Error deleting uploaded file:', unlinkErr);
            });
        }
        return res.status(500).json({ success: false, message: 'Error saving profile picture.' });
    }

    // Check if no rows were affected (user not found or not owned by user)
    if (data && data.length === 0) {
        // For Vercel deployments using Vercel Blob, we don't need to delete local files
        // For local development, delete the file
        if (!process.env.VERCEL) {
            let filePath = req.file.path;
            if (req.file.path && req.file.path.startsWith('/tmp')) {
                const imagesDir = getImagesDirectory(__dirname);
                filePath = path.join(imagesDir, req.file.filename);
            }
            fs.unlink(filePath, (unlinkErr) => {
                if (unlinkErr) console.error('Error deleting uploaded file:', unlinkErr);
            });
        }
        return res.status(404).json({ success: false, message: 'User not found or you do not have permission to update profile picture.' });
    }

    // For Vercel Blob URLs, return as-is since they're already full URLs
    // For local images, construct full URL
    const fullProfilePicturePath = profilepicturePath.startsWith('http') ? 
        profilepicturePath : 
        `${BASE_URL}${profilepicturePath.replace(/\\/g, '/')}`;
    res.json({ success: true, message: 'Profile picture updated successfully!', profilepicture: fullProfilePicturePath });
};

exports.getProfilePicture = async (req, res) => {
    const userId = req.user.id;

    const { data: users, error } = await supabase
        .from('users')
        .select('profilePicture')
        .eq('id', userId);

    if (error) {
        console.error(error);
        return res.status(500).json({ success: false, message: 'An error occurred.' });
    }

    if (users.length > 0) {
        let profilepicture = users[0].profilePicture;
        // For Vercel Blob URLs, return as-is since they're already full URLs
        // For local images, return relative path
        if (profilepicture && !profilepicture.startsWith('http')) {
            profilepicture = profilepicture.replace(/\\/g, '/');
        }
        res.json({ success: true, profilepicture: profilepicture });
    } else {
        res.status(404).json({ success: false, message: 'User not found.' });
    }
};

exports.verifyCurrentPassword = async (req, res) => {
    const { currentPassword } = req.body;
    const userId = req.user.id;

    if (!currentPassword) {
        return res.status(400).json({ success: false, message: 'Current password is required.' });
    }

    const { data: users, error } = await supabase
        .from('users')
        .select('password')
        .eq('id', userId);

    if (error) {
        console.error(error);
        return res.status(500).json({ success: false, message: 'An error occurred.' });
    }

    if (users.length === 0) {
        return res.status(404).json({ success: false, message: 'User not found.' });
    }

    const hashedPassword = users[0].password;
    bcrypt.compare(currentPassword, hashedPassword, (compareErr, isMatch) => {
        if (compareErr) {
            console.error(compareErr);
            return res.status(500).json({ success: false, message: 'An error occurred during password comparison.' });
        }
        if (isMatch) {
            res.json({ success: true, message: 'Current password matches.' });
        } else {
            res.status(401).json({ success: false, message: 'Current password does not match.' });
        }
    });
};

exports.changePassword = async (req, res) => {
    const { currentPassword, newPassword, confirmNewPassword } = req.body;
    const userId = req.user.id;
    const userEmail = req.user.email;

    console.log('/change-password: Request received for user ID:', userId, 'Email:', userEmail);

    if (!currentPassword || !newPassword || !confirmNewPassword) {
        return res.status(400).json({ success: false, message: 'All password fields are required.' });
    }

    if (newPassword !== confirmNewPassword) {
        return res.status(400).json({ success: false, message: 'New password and confirm password do not match.' });
    }

    const { data: users, error: checkError } = await supabase
        .from('users')
        .select('password')
        .eq('id', userId);

    if (checkError) {
        console.error('/change-password: Error verifying current password from DB:', checkError);
        return res.status(500).json({ success: false, message: 'An error occurred while verifying current password.' });
    }

    if (users.length === 0) {
        console.log('/change-password: User not found for ID:', userId);
        return res.status(404).json({ success: false, message: 'User not found.' });
    }

    const hashedPassword = users[0].password;
    bcrypt.compare(currentPassword, hashedPassword, async (compareErr, isMatch) => {
        if (compareErr) {
            console.error('/change-password: Error comparing current password:', compareErr);
            return res.status(500).json({ success: false, message: 'An error occurred during current password verification.' });
        }
        if (!isMatch) {
            console.log('/change-password: Invalid current password for user ID:', userId);
            return res.status(401).json({ success: false, message: 'Invalid current password.' });
        }

        bcrypt.hash(newPassword, 10, async (hashErr, newHashedPassword) => {
            if (hashErr) {
                console.error('/change-password: Error hashing new password:', hashErr);
                return res.status(500).json({ success: false, message: 'An error occurred during new password hashing.' });
            }

            const { data, error: updateError } = await supabase
                .from('users')
                .update({ password: newHashedPassword })
                .eq('id', userId);

            if (updateError) {
                console.error('/change-password: Error updating password in DB:', updateError);
                return res.status(500).json({ success: false, message: 'An error occurred while changing password.' });
            }

            console.log('/change-password: Password updated in DB for user ID:', userId);

            const expiresIn = '1h';
            const newAccessToken = jwt.sign({ id: userId, email: userEmail }, JWT_SECRET, { expiresIn });
            console.log('/change-password: Generated new token:', newAccessToken.substring(0, 10) + '...');

            const { error: tokenUpdateError } = await supabase
                .from('users')
                .update({ token: newAccessToken })
                .eq('id', userId);

            if (tokenUpdateError) {
                console.error('/change-password: Error updating token after password change:', tokenUpdateError);
                return res.status(500).json({ success: false, message: 'An error occurred while updating session.' });
            }

            console.log('/change-password: Token updated in DB for user ID:', userId);
            res.json({ success: true, message: 'Password changed successfully!', token: newAccessToken });
        });
    });
};

File: backend/controllers/user_controller.py
from flask import request, jsonify, g
import bcrypt
import jwt
import datetime
from utils.supabase_client import supabase
from utils.supabase_storage import upload_file_to_supabase, delete_file_from_supabase
from config import Config

def get_user_info():
    user_id = g.user['id']
    print(f'getUserInfo: Fetching user info for user ID: {user_id}')

    try:
        response = supabase.table('users').select('id, firstname, middlename, lastname, email, profilepicture').eq('id', user_id).execute()
        users = response.data

        if users and len(users) > 0:
            user = users[0]
            if not user.get('profilepicture'):
                user['profilepicture'] = 'https://nttadnyxpbuwuhgtpvjh.supabase.co/storage/v1/object/public/images/default-profile.png'
            
            return jsonify({'success': True, 'user': user})
        else:
            return jsonify({'success': False, 'message': 'User not found.'}), 404
    except Exception as e:
        print(f'Error in getUserInfo: {e}')
        return jsonify({'success': False, 'message': 'An error occurred while fetching user information.'}), 500

def update_user_info(id):
    user_id = g.user['id']
    if str(user_id) != str(id):
        return jsonify({'success': False, 'message': 'Unauthorized to update this user.'}), 403

    data = request.get_json()
    firstname = data.get('firstname')
    middlename = data.get('middlename')
    lastname = data.get('lastname')
    email = data.get('email')

    if not all([firstname, lastname, email]):
        return jsonify({'success': False, 'message': 'First name, last name, and email are required.'}), 400

    try:
        supabase.table('users').update({
            'firstname': firstname,
            'middlename': middlename,
            'lastname': lastname,
            'email': email
        }).eq('id', user_id).execute()
        
        return jsonify({'success': True, 'message': 'Account information updated successfully!'})
    except Exception as e:
        print(f'Error updating user info: {e}')
        return jsonify({'success': False, 'message': 'Error updating user information.'}), 500

def upload_profile_picture():
    user_id = g.user['id']
    if 'profilePicture' not in request.files:
        return jsonify({'success': False, 'message': 'No file uploaded.'}), 400

    file = request.files['profilePicture']
    if file.filename == '':
        return jsonify({'success': False, 'message': 'No file selected.'}), 400

    try:
        # Get current profile picture
        response = supabase.table('users').select('profilepicture').eq('id', user_id).execute()
        if not response.data:
            return jsonify({'success': False, 'message': 'User not found.'}), 404
            
        current_profile_picture = response.data[0].get('profilepicture')
        
        # Upload new file
        file_content = file.read()
        file_name = f"profile-pictures/{file.filename}"
        
        upload_result = upload_file_to_supabase(file_content, file_name, 'images')
        
        if upload_result['error']:
            print(f"Error uploading: {upload_result['error']}")
            return jsonify({'success': False, 'message': 'Failed to upload profile picture.'}), 500
            
        profile_picture_path = upload_result['publicUrl']
        
        # Delete old file if it exists and is not default
        if current_profile_picture and 'supabase.co/storage' in current_profile_picture and 'default-profile.png' not in current_profile_picture:
            try:
                # Extract path from URL
                # URL: https://.../storage/v1/object/public/images/path/to/file
                parts = current_profile_picture.split('/public/images/')
                if len(parts) > 1:
                    old_file_path = parts[1]
                    delete_file_from_supabase(old_file_path, 'images')
            except Exception as e:
                print(f"Error deleting old file: {e}")

        # Update user record
        supabase.table('users').update({'profilepicture': profile_picture_path}).eq('id', user_id).execute()
        
        return jsonify({'success': True, 'message': 'Profile picture updated successfully!', 'profilepicture': profile_picture_path})

    except Exception as e:
        print(f'Error in uploadProfilePicture: {e}')
        return jsonify({'success': False, 'message': 'Error saving profile picture.'}), 500

def get_profile_picture():
    user_id = g.user['id']
    try:
        response = supabase.table('users').select('profilepicture').eq('id', user_id).execute()
        if response.data:
            profile_picture = response.data[0].get('profilepicture')
            if not profile_picture:
                profile_picture = 'https://nttadnyxpbuwuhgtpvjh.supabase.co/storage/v1/object/public/images/default-profile.png'
            return jsonify({'success': True, 'profilepicture': profile_picture})
        else:
            return jsonify({'success': False, 'message': 'User not found.'}), 404
    except Exception as e:
        print(f'Error in getProfilePicture: {e}')
        return jsonify({'success': False, 'message': 'An unexpected error occurred.'}), 500

def verify_current_password():
    user_id = g.user['id']
    data = request.get_json()
    current_password = data.get('currentPassword')

    if not current_password:
        return jsonify({'success': False, 'message': 'Current password is required.'}), 400

    try:
        response = supabase.table('users').select('password').eq('id', user_id).execute()
        if not response.data:
            return jsonify({'success': False, 'message': 'User not found.'}), 404
            
        hashed_password = response.data[0]['password']
        if bcrypt.checkpw(current_password.encode('utf-8'), hashed_password.encode('utf-8')):
            return jsonify({'success': True, 'message': 'Current password matches.'})
        else:
            return jsonify({'success': False, 'message': 'Current password does not match.'}), 401
    except Exception as e:
        print(f'Error verifying password: {e}')
        return jsonify({'success': False, 'message': 'An error occurred.'}), 500

def change_password():
    user_id = g.user['id']
    user_email = g.user['email']
    data = request.get_json()
    current_password = data.get('currentPassword')
    new_password = data.get('newPassword')
    confirm_new_password = data.get('confirmNewPassword')

    if not all([current_password, new_password, confirm_new_password]):
        return jsonify({'success': False, 'message': 'All password fields are required.'}), 400

    if new_password != confirm_new_password:
        return jsonify({'success': False, 'message': 'New password and confirm password do not match.'}), 400

    try:
        # Verify current password
        response = supabase.table('users').select('password').eq('id', user_id).execute()
        if not response.data:
            return jsonify({'success': False, 'message': 'User not found.'}), 404
            
        hashed_password = response.data[0]['password']
        if not bcrypt.checkpw(current_password.encode('utf-8'), hashed_password.encode('utf-8')):
            return jsonify({'success': False, 'message': 'Invalid current password.'}), 401

        # Hash new password
        new_hashed_password = bcrypt.hashpw(new_password.encode('utf-8'), bcrypt.gensalt()).decode('utf-8')
        
        # Update password
        supabase.table('users').update({'password': new_hashed_password}).eq('id', user_id).execute()
        
        # Generate new token
        new_access_token = jwt.encode({
            'id': user_id,
            'email': user_email,
            'exp': datetime.datetime.utcnow() + datetime.timedelta(hours=1)
        }, Config.JWT_SECRET, algorithm="HS256")
        
        supabase.table('users').update({'token': new_access_token}).eq('id', user_id).execute()
        
        return jsonify({'success': True, 'message': 'Password changed successfully!', 'token': new_access_token})

    except Exception as e:
        print(f'Error changing password: {e}')
        return jsonify({'success': False, 'message': 'An error occurred while changing password.'}), 500
